pkg/apis/resource/validation/validation.go

```diff
@@ -621,27 +661,77 @@ func validateResourcePool(pool resource.ResourcePool, fldPath *field.Path) field
-	allErrs = append(allErrs, validateMap(device.Capacity, -1, maxKeyLen, validateQualifiedName, validateDeviceCapacity, fldPath.Child("capacity"))...)
+	allErrs = append(allErrs, validateMap(device.Attributes, -1, attributeAndCapacityMaxKeyLength, validateQualifiedName, validateDeviceAttribute, fldPath.Child("attributes"))...)
+	allErrs = append(allErrs, validateMap(device.Capacity, -1, attributeAndCapacityMaxKeyLength, validateQualifiedName, validateDeviceCapacity, fldPath.Child("capacity"))...)
 	if combinedLen, max := len(device.Attributes)+len(device.Capacity), resource.ResourceSliceMaxAttributesAndCapacitiesPerDevice; combinedLen > max {
 		allErrs = append(allErrs, field.Invalid(fldPath, combinedLen, fmt.Sprintf("the total number of attributes and capacities must not exceed %d", max)))
 	}
+
+	allErrs = append(allErrs, validateSet(device.ConsumesCapacity, resource.ResourceSliceMaxDeviceCapacityConsumptions,
+		func(deviceCapacityConsumption resource.DeviceCapacityConsumption, fldPath *field.Path) field.ErrorList {
+			return validateDeviceCapacityConsumption(deviceCapacityConsumption, fldPath, capacityPoolNames)
+		},
+		func(deviceCapacityConsumption resource.DeviceCapacityConsumption) (string, string) {
+			return deviceCapacityConsumption.CapacityPool, "capacityPool"
+		}, fldPath.Child("consumesCapacity"))...)
+
+	if perDeviceNodeSelection {
+		setFields := make([]string, 0, 3)
+		if device.NodeName != "" {
+			setFields = append(setFields, "`nodeName`")
+			allErrs = append(allErrs, validateNodeName(device.NodeName, fldPath.Child("nodeName"))...)
+		}
+		if device.NodeSelector != nil {
+			setFields = append(setFields, "`nodeSelector`")
+			allErrs = append(allErrs, corevalidation.ValidateNodeSelector(device.NodeSelector, false, fldPath.Child("nodeSelector"))...)
+		}
+		if device.AllNodes {
+			setFields = append(setFields, "`allNodes`")
+		}
+		switch len(setFields) {
+		case 0:
+			allErrs = append(allErrs, field.Required(fldPath, "exactly one of `nodeName`, `nodeSelector`, or `allNodes` is required when `perDeviceNodeSelection` is set in the ResourceSlice spec"))
+		case 1:
+		default:
+			allErrs = append(allErrs, field.Invalid(fldPath, fmt.Sprintf("{%s}", strings.Join(setFields, ", ")), "exactly one of `nodeName`, `nodeSelector`, or `allNodes` is required when `perDeviceNodeSelection` is set in the ResourceSlice spec"))
+		}
+	} else if device.NodeName != "" || device.NodeSelector != nil || device.AllNodes {
+		allErrs = append(allErrs, field.Invalid(fldPath, nil, "`nodeName`, `nodeSelector` and `allNodes` can only be set if `perDeviceNodeSelection` is set in the ResourceSlice spec"))
+	}
+
+	return allErrs
+}
+
+func validateDeviceCapacityConsumption(deviceCapacityConsumption resource.DeviceCapacityConsumption, fldPath *field.Path, capacityPoolNames sets.Set[string]) field.ErrorList {
+	var allErrs field.ErrorList
+
+	if len(deviceCapacityConsumption.CapacityPool) == 0 {
```

Comment:
Currently we only validate that the DeviceCapacityConsumption entries references an existing CapacityPool. We could also validate that all the capacities referenced also exists in the pool. And if we want to go even further, we could verify that the requested capacities doesn't exceed what is available in the pool (we can't guarantee that the device can be allocated if it less than what is in the pool, but if it is larger, we know it can never be allocated).

But I'm not sure if we want to validate to this level of detail. This also gets more complicated when we introduce the mixins... Any thoughts @pohly @johnbelamaric

Reply:
Thanks for the discussion. I have added the within object reference validation. And I didn't see other API doing the across capacities reference check so I am gonna assume this is enough as of now. Let me know if further validation is needed^^